include/rmm/mr/device/tracking_resource_adaptor.hpp

```diff
@@ -199,9 +253,39 @@ class tracking_resource_adaptor final : public device_memory_resource {
     upstream_->deallocate(p, bytes, stream);
     {
       write_lock_t lock(mtx_);
-      allocations_.erase(p);
+
+      const auto found = allocations_.find(p);
+
+      // Ensure the allocation is found and the number of bytes match
+      if (found == allocations_.end()) {
+        // Don't throw but log an error. Throwing in a descructor (or any noexcept) will call
+        // std::terminate
+        RMM_LOG_ERROR(static_cast<std::ostringstream&&>(
+                        std::ostringstream()
+                        << "Deallocating a pointer that was not tracked. Ptr:" << p
+                        << ", Bytes: " << bytes << ", Alloc Size: " << this->allocations_.size())
+                        .str());
+      } else {
+        allocations_.erase(found);
+
+        auto allocated_bytes = found->second.allocation_size;
+
+        if (allocated_bytes != bytes) {
+          // Don't throw but log an error. Throwing in a descructor (or any noexcept) will call
+          // std::terminate
+          RMM_LOG_ERROR(static_cast<std::ostringstream&&>(std::ostringstream()
+                                                          << "Alloc bytes (" << allocated_bytes
+                                                          << ") and Dealloc bytes (" << bytes
+                                                          << ") do not match")
+                          .str());
+          bytes = allocated_bytes;
+        }
+      }
+
+      // Decrement the current allocated counts.
+      allocation_counts_.current_bytes -= bytes;
+      allocation_counts_.current_count -= 1;
```

Comment:
This should also be encapsulated in a member function.

Reply:
Moved to a member function